README.md
![ColorizeSwift](Assets/logo.png)

[![Swift](https://img.shields.io/badge/language-Swift-orange.svg?style=flat)](https://developer.apple.com/swift/)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/mtynior/ColorizeSwift/blob/master/LICENSE.md) 

Terminal string styling for Swift.

## Integration

#### Swift Package Manager (SPM)
You can use The [Swift Package Manager](https://swift.org/package-manager/) to install ColorizeSwift by adding it to your Package.swift file:

```swift
// swift-tools-version:4.0
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "MyLibrary",
    products: [
        .library(name: "MyLibrary", targets: ["MyLibrary"]),
    ],
    dependencies: [
         .package(url: "https://github.com/mtynior/ColorizeSwift.git", from: "1.5.0"),
    ],
    targets: [
        .target(name: "MyLibrary", dependencies: ["ColorizeSwift"]),
        .testTarget(name: "MyLibraryTests", dependencies: ["MyLibrary", "ColorizeSwift"]),
    ]
)
```

#### Manually
You can also manually add ColorizeSwift to you project:

1. Download `ColorizeSwift.swift` file,
2. Drag `ColorizeSwift.swift` into you project's tree.


#### CocoaPods

#### From version `1.5` ColorizeSwift does not support CocoaPods!
Please user SPM or add `ColorizeSwift.swift` file  manually to your project. 
If you use CocoaPods, you can still use version `1.2`.  **At the moment, latest version`1.5`, only removed support for CocoaPods and has the same functionality as `1.2`, but it might change in the future.** 

You can use [CocoaPods](http://cocoapods.org/) to install `ColorizeSwift` by adding it to your `Podfile`:

```ruby
platform :ios, '9.0'
use_frameworks!

target 'MyApp' do
	pod 'ColorizeSwift'
end
```
Run `pods install` to intagrate pods with your project.


## Example
You can run sample application:

1. Open Terminal and go to `Example` folder.
2. Run `./build.sh` script to build sample application.
3. Run `./example pacman` to launch sample.

Available samples:

1. `styles` - prints available styles 

	![Example - Styles](Assets/styles.png)

2. `f1` - prints F1 cars

	![Example - F1](Assets/f1.png)
	
3. `pacman` - prints Pacman

	![Example - Pacman](Assets/pacman.png)

4. `mario` - prints mario

	![Example - Mario](Assets/mario.png)

## Usage
```swift
print("Normal")
print("Bold".bold())
print("Dim".dim())
print("Italic".italic())
print("Underline".underline())
print("Blink".blink())
print("Reverse".reverse())
print("hidden".hidden())
print("strikethrough".strikethrough())
print("Red".red())
print("On yellow".onYellow())
print("256 foreground".foregroundColor(.orange1))
print("226 background".backgroundColor(.orange1))
print("Awful combination".colorize(.yellow, background: .red))
    
let nested = "with a blue substring".blue().underline()
print("A bold, green line \(nested) that becomes bold and green again".green().bold())
```

## Styles

### Modifiers

- `bold()`
- `dim()`
- `italic()` *(not widely supported)*
- `underline()`
- `reverse()`
- `hidden()`
- `strikethrough()` *(not widely supported)*
- `reset()`

### Foreground colors

- `black()`
- `red()`
- `green()`
- `yellow()`
- `blue()`
- `magenta()`
- `cyan()`
- `lightGray()`
- `darkGray()`
- `lightRed()`
- `lightGreen()`
- `lightYellow()`
- `lightBlue()`
- `lightMagenta()`
- `lightCyan()`
- `white()`

### Background colors

- `onBlack()`
- `onRed()`
- `onGreen()`
- `onYellow()`
- `onBlue()`
- `onMagenta()`
- `onCyan()`
- `onLightGray()`
- `onDarkGray()`
- `onLightRed()`
- `onLightGreen()`
- `onLightYellow()`
- `onLightBlue()`
- `onLightMagenta()`
- `onLightCyan()`
- `onWhite()`


## 256-colors
You can also use 256 colors, but keep in mind that not all Terminal clients support them.

- `foregroundColor(color: TerminalColor)`
- `backgroundColor(color: TerminalColor)`
- `colorize(foreground: TerminalColor, background: TerminalColor)`

#### Available colors

You can access 256 colors using `TerminalColor` enumeration.

![256 Colors](Assets/256Colors.png)

## Escape codes
Sometimes you only need the `open code` for a modifier. You can access them using `TerminalStyle` enum:

```swift
TerminalStyle.bold.open // "\u{001B}[1m"
TerminalStyle.bold.close  // "\u{001B}[22m"
```

For 256 colors use:

```swift
TerminalColor.red.foregroundStyleCode().open \\"\u{001B}[38;5;9m"
TerminalColor.red.backgroundStyleCode().open \\"\u{001B}[48;5;9m"
```

## Uncolorize

To get string without any colorization use `uncolorized()` method:
```swift
let styledString = "Awful combination".colorize(.yellow, background: .red) // \u{001B}[48;5;9m\u{001B}[38;5;11mAwful combination\u{001B}[0m\u{001B}[48;5;9m\u{001B}[0m
let withoutStyles = styledString.uncolorized() // "Awful combination"
```

## Disabling colorization

Colorization can be disabled globally:

```swift
String.isColorizationEnabled = false // Default: true
```

You can use this to support a command line option (`./example --no-color`):
```swift
String.isColorizationEnabled = !CommandLine.arguments.contains("--no-color")
```
To disable colorization when program is running in a pipe (`./example | wc`) or when not writing to `stdout` (`./example > output.txt`):
```swift
String.isColorizationEnabled = (isatty(fileno(stdout)) == 1)
```


## Credits

- [Mike Smiley](https://github.com/msmiley)
- [Theo Lubert](https://github.com/theo-lubert) 
- [Noah Bass](https://github.com/noahbass)
- [Hristo Staykov](https://github.com/hristost)

## License

ColorizeSwift is released under the MIT license. See LICENSE for details.


CHANGELOG.md
# Change Log

#### 1.x Releases

- `1.6.x` Releases - [1.6.0](#160)

- `1.5.x` Releases - [1.5.0](#150)

- `1.2.x` Releases - [1.2.0](#120)

- `1.1.x` Releases - [1.1.0](#110)

- `1.0.x` Releases - [1.0.0](#100) | [1.0.1](#101) | [1.0.2](#102) | [1.0.3](#103) 

---
## [1.6.0](https://github.com/mtynior/ColorizeSwift/releases/tag/1.6.0)
Released on 2020-08-09.

#### Added
- `uncolorized()`  method, which removes any colorization


## [1.5.0](https://github.com/mtynior/ColorizeSwift/releases/tag/1.5.0)
Released on 2020-03-03.

#### Updated
- Source code to Swift 5.0 and new Swift Package Manager    

#### Removed
Support for CocoaPods.

## [1.2.0](https://github.com/mtynior/ColorizeSwift/releases/tag/1.2.0)
Released on 2017-09-05.

#### Added
- Ability to disable colorization with `String.isColorizationEnnabled`, thanks to @theo-lubert.

## [1.1.0](https://github.com/mtynior/ColorizeSwift/releases/tag/1.1.0)
Released on 2016-11-01.

#### Updated
- Source code to Swift 3.0	
- Project to Xcode 8

## [1.0.3](https://github.com/mtynior/ColorizeSwift/releases/tag/1.0.3)
Released on 2016-06-29.

#### Fixed
- Includes/import in header file

## [1.0.1](https://github.com/mtynior/ColorizeSwift/releases/tag/1.0.1)
Released on 2016-04-02.

#### Updated
- Social media URL in `podspec`	

## [1.0.0](https://github.com/mtynior/ColorizeSwift/releases/tag/1.0.0)
Released on 2016-04-01.

#### Added
- Initial release of ColorizeSwift


Example/main.swift
//#!cat ../Source/ColorizeSwift.swift main.swift | swift -

//
//  main.swift
//  Colorize
//
//  Created by Michał Tynior on 07/01/16.
//  Copyright © 2016 Michał Tynior. All rights reserved.
//

func printStyles() {
    print("Normal")
    print("Bold".bold())
    print("Dim".dim())
    print("Italic".italic())
    print("Underline".underline())
    print("Blink".blink())
    print("Reverse".reverse())
    print("Hidden".hidden())
    print("strikethrough".strikethrough())
    print("Red".red())
    print("On yellow".onYellow())
    print("256 foreground".foregroundColor(.orange1))
    print("226 background".backgroundColor(.orange1))
    print("Awful combination".colorize(.yellow, background: .red))
    
    let nested = "with a blue, underline substring".blue().underline()
    print("A bold, green line \(nested) that becomes bold and green again".green().bold())
}

func printF1() {
    
    let car1 = TerminalColor.green.foregroundStyleCode().open
    let car2 = TerminalColor.red.foregroundStyleCode().open
    let car3 = TerminalColor.mediumPurple.foregroundStyleCode().open
    let car4 = TerminalColor.yellow1.foregroundStyleCode().open
    
    print("\n\n")
    print("██  ██  ██  ██  ██  ██  ██  ██  ██  ██".white())
    print("  ██  ██  ██  ██  ██  ██  ██  ██  ██  ".white())
    
    print("\n\n")
    print(car1 + "      ▄▄           " )
    print(car1 + "   ██▬██▬██        " + car2 + "      ▄▄      ")
    print(car1 + "     ████          " + car2 + "   ██▬██▬██   ")
    print(car1 + "   ▄██  ██▄        " + car2 + "     ████     ")
    print(car1 + " ▄▄▄▀█  █▀▄▄▄      " + car2 + "   ▄██  ██▄   ")
    print(car1 + " ███▀████▀███      " + car2 + " ▄▄▄▀█  █▀▄▄▄ ")
    print(car1 + "     ▀  ▀          " + car2 + " ███▀████▀███ ")
    print(car1 + "                   " + car2 + "     ▀  ▀     ")
    print(car1 + "\n\n")
    print(car3 + "      ▄▄           " )
    print(car3 + "   ██▬██▬██        " + car4 + "      ▄▄      ")
    print(car3 + "     ████          " + car4 + "   ██▬██▬██   ")
    print(car3 + "   ▄██  ██▄        " + car4 + "     ████     ")
    print(car3 + " ▄▄▄▀█  █▀▄▄▄      " + car4 + "   ▄██  ██▄   ")
    print(car3 + " ███▀████▀███      " + car4 + " ▄▄▄▀█  █▀▄▄▄ ")
    print(car3 + "     ▀  ▀          " + car4 + " ███▀████▀███ ")
    print(car3 + "                   " + car4 + "     ▀  ▀     ")
    print("\n\n")
    
}

func printPacman() {
    
    // inspired by https://github.com/jaykru/colorscripts
    
    let pacmanColor = TerminalColor.yellow.foregroundStyleCode().open
    let blinkyColor = TerminalColor.red.foregroundStyleCode().open
    let pinkyColor = TerminalColor.pink1.foregroundStyleCode().open
    let inkyColor = TerminalColor.blue.foregroundStyleCode().open
    let clydeColor = TerminalColor.gold1.foregroundStyleCode().open
    let white = TerminalColor.white.foregroundStyleCode().open
    
    print("\n\n")
    print("\(pacmanColor)  ▄███████▄       \(blinkyColor)  ▄██████▄       \(pinkyColor)  ▄██████▄       \(inkyColor)  ▄██████▄       \(clydeColor)  ▄██████▄  ")
    print("\(pacmanColor)▄█████████▀▀      \(blinkyColor)▄\(white)█▀█\(blinkyColor)██\(white)█▀█\(blinkyColor)██▄     \(pinkyColor)▄\(white)█▀█\(pinkyColor)██\(white)█▀█\(pinkyColor)██▄     \(inkyColor)▄\(white)█▀█\(inkyColor)██\(white)█▀█\(inkyColor)██▄     \(clydeColor)▄\(white)█▀█\(clydeColor)██\(white)█▀█\(clydeColor)██▄")
    print("\(pacmanColor)███████▀          \(blinkyColor)█\(white)▄▄█\(blinkyColor)██\(white)▄▄█\(blinkyColor)███     \(pinkyColor)█\(white)▄▄█\(pinkyColor)██\(white)▄▄█\(pinkyColor)███     \(inkyColor)█\(white)▄▄█\(inkyColor)██\(white)▄▄█\(inkyColor)███     \(clydeColor)█\(white)▄▄█\(clydeColor)██\(white)▄▄█\(clydeColor)███")
    print("\(pacmanColor)███████▄          \(blinkyColor)████████████     \(pinkyColor)████████████     \(inkyColor)████████████     \(clydeColor)████████████")
    print("\(pacmanColor)▀█████████▄▄      \(blinkyColor)██▀██▀▀██▀██     \(pinkyColor)██▀██▀▀██▀██     \(inkyColor)██▀██▀▀██▀██     \(clydeColor)██▀██▀▀██▀██")
    print("\(pacmanColor)  ▀███████▀       \(blinkyColor)▀   ▀  ▀   ▀     \(pinkyColor)▀   ▀  ▀   ▀     \(inkyColor)▀   ▀  ▀   ▀     \(clydeColor)▀   ▀  ▀   ▀")
    print("\n\n")
}

func printMario() {
    
    let red = TerminalColor.red.foregroundStyleCode().open
    let brown = TerminalColor.darkRed.foregroundStyleCode().open
    let skin = TerminalColor.sandyBrown.foregroundStyleCode().open
    let eye = TerminalColor.black.foregroundStyleCode().open
    let trousers = TerminalColor.deepSkyBlue4_2.foregroundStyleCode().open
    let button = TerminalColor.yellow.foregroundStyleCode().open
    
    print("\n\n")
    print("\(red)    █████    ")
    print("\(red)   █████████ ")
    print("\(brown)   ███\(skin)██\(eye)█\(skin)█   ")
    print("\(brown)  █\(skin)█\(brown)█\(skin)███\(eye)█\(skin)███ ")
    print("\(brown)  █\(skin)█\(brown)██\(skin)████\(brown)█\(skin)██")
    print("\(brown)  ██\(skin)████\(brown)████ ")
    print("\(skin)     ██████  ")
    print("\(red)    ██\(trousers)█\(red)███    ")
    print("\(red)   ███\(trousers)█\(red)██\(trousers)█\(red)███ ")
    print("\(red)  ████\(trousers)████\(red)████")
    print("\(skin)  ██\(red)█\(trousers)█\(button)█\(trousers)██\(button)█\(trousers)█\(red)█\(skin)██")
    print("\(skin)  ███\(trousers)██████\(skin)███")
    print("\(skin)  ██\(trousers)████████\(skin)██")
    print("\(trousers)    ███  ███  ")
    print("\(brown)   ███    ███ ")
    print("\(brown)  ████    ████")
    print("\n\n")
    
}

func printInfo() {
    print("Available examples:\n")
    print("   styles    Print available styles.")
    print("   f1        Print F1 cars.")
    print("   pacman    Print Pacman")
    print("   mario     Print Mario")
    print("\n")
}

if CommandLine.arguments.count != 2 {
    printInfo()
}
else {
    let argument = CommandLine.arguments[1]

    switch (argument.lowercased()) {
        case "styles": printStyles()
        case "f1": printF1()
        case "pacman": printPacman()
        case "mario": printMario()
        default:
            print("Unrecognized command `\(argument)`.")
            printInfo()
    }

}
